### PR TITLE
add scons fmt check for travis linux build

### LINKED ISSUE #309: Add scons fmt check to Travis builds
We have `scons fmt` command that formats the code according to our code style. It adds necessary copyright headers and runs clang-format.

It would be nice to check that the code is formatted in out Travis build, i.e. run `scons fmt` and ensure that it did not modify any source files (using git status I guess). If it did, fail the build.

The check can be added to the "Linux build checks" section: https://github.com/roc-project/roc/blob/4ebf1052c2d2048a2a4cb897a002c21d180e3bdc/.travis.yml#L60-L72

### PR DESCRIPTION
**Pull request by fusuiyi123**
Fixes #309

### CODE REVIEW

**Comment by gavv** [comment]
Cool!

Please re-target PR to develop branch, as stated in [CONTRIBUTING.md](https://github.com/roc-project/roc/blob/master/CONTRIBUTING.md).

Also, let's move the check (scons + git) to a script, say, `scripts/travis/linux-build-checks/code-formatting.sh`.

Also, it would be nice to produce a meaningful error message if the check fails. Something like: "the code is not formatted, please run scons fmt, using clang-format version x.y.z". And probably to would be useful to print a diff to see what exactly is not formatted.

And we should run the script inside a container with scons and clang-format installed. You can find the list of available images, and documentation on them, [here](https://roc-project.github.io/roc/docs/development/continuous_integration.html). I think we can use `travis-checks` image. If it doesn't have all necessary tools, you can add them by issuing a pull request to [this repo](https://github.com/roc-project/dockerfiles). After merging the PR, Docker Hub will automatically rebuild the image from the updated Dockerfile.

What clang-format version to you use? I'm using 9.0.1 and it does not modify `of_decoder.cpp` when I run `scons fmt`.

**Comment by fusuiyi123** [comment]
thanks! this is my version, `clang-format version 9.0.0 (tags/google/stable/2019-05-14)` let me rebase to development.  Actually I tried moving scons and git to script earlier but it will give `git command not found` error, I will try and search again.

**Comment by fusuiyi123** [comment]
https://github.com/roc-project/dockerfiles/pull/2

**Comment by gavv** [comment]
> thanks! this is my version, clang-format version 9.0.0 (tags/google/stable/2019-05-14) let me rebase to development

Ah, got it, `of_decoder.cpp` was modified because it was not formatted in master. OK.

File: scripts/travis/linux-build-checks/check-fmt.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#! /bin/bash
+set -euxo pipefail
```
Comment (gavv) [inline scripts/travis/linux-build-checks/check-fmt.sh]: Let's remove `-x`, for cleaner output.

File: scripts/travis/linux-build-checks/check-fmt.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+scons -Q fmt
+
+diff=$(git status --porcelain)
+
+if [[ $diff ]]; then
+    echo $diff
+    echo "format check FAILED"
+    echo "please install clang-format and run scons -Q fmt"
+    exit 1
```
Comment (gavv) [inline scripts/travis/linux-build-checks/check-fmt.sh]: I suggest a small improvement:

```
if [[ $diff ]]; then
    echo
    echo "format check FAILED"
    echo "please install clang-format and run scons -Q fmt"
    echo
    GIT_PAGER= git diff
    exit 1
```

The output will be a bit more convenient IMHO.

image

**Comment by fusuiyi123** [comment]
cool, applied your comments

**Comment by gavv** [comment]
We should also add `docker pull rocproject/travis-checks` to `before_install` section of `Linux build checks` job.

**Comment by gavv** [comment]
dockerfiles PR is merged. Docker hub will rebuild the image in a few hours. https://hub.docker.com/r/rocproject/travis-checks

**Comment by gavv** [comment]
Thanks!

**Pull request merged**